tests/testSdssShapePsf.py

```diff
@@ -0,0 +1,128 @@
+        # third source is extended
+        self.dataset.addSource(flux=1E5, centroid=afwGeom.Point2D(149.9, 50.3),
+                               shape=afwGeom.ellipses.Quadrupole(8, 9, 3))
+        self.config = self.makeSingleFrameMeasurementConfig("base_SdssShape")
+
+    def tearDown(self):
+        del self.bbox
+        del self.dataset
+        del self.config
+
+    def _computeVaryingPsf(self):
+        """Compute a varying PSF as a linear combination of PCA (== Karhunen-Loeve) basis functions
+
+        We simply desire a psf that is not constant across the image, so the precise choice of
+        parameters (e.g., sigmas, setSpatialParameters) are not crucial.
+        """
+        kernelSize = 31
+        sigma1 = 1.75
+        sigma2 = 2.0*sigma1
+        basisKernelList = afwMath.KernelList()
+        for sigma in (sigma1, sigma2):
+            basisKernel = afwMath.AnalyticKernel(kernelSize, kernelSize,
+                                                 afwMath.GaussianFunction2D(sigma, sigma))
+            basisImage = afwImage.ImageD(basisKernel.getDimensions())
+            basisKernel.computeImage(basisImage, True)
+            basisImage /= np.sum(basisImage.getArray())
+            if sigma == sigma1:
+                basisImage0 = basisImage
+            else:
+                basisImage -= basisImage0
+            basisKernelList.append(afwMath.FixedKernel(basisImage))
+
+        order = 1
+        spFunc = afwMath.PolynomialFunction2D(order)
+        exactKernel = afwMath.LinearCombinationKernel(basisKernelList, spFunc)
+        exactKernel.setSpatialParameters([[1.0, 0, 0], [0.0, 0.5E-2, 0.2E-2]])
+        exactPsf = measAlg.PcaPsf(exactKernel)
+
+        return exactPsf
+
+    def _runMeasurementTask(self, psf=None):
+        task = self.makeSingleFrameMeasurementTask("base_SdssShape", config=self.config)
+        exposure, catalog = self.dataset.realize(10.0, task.schema)
+        if psf:
+            exposure.setPsf(psf)
+        task.run(exposure, catalog)
+        return exposure, catalog
+
+    def _checkPsfShape(self, result, psfResult, psfTruth):
+        self.assertClose(psfResult.getIxx(), psfTruth.getIxx(), rtol=1E-4)
+        self.assertClose(psfResult.getIyy(), psfTruth.getIyy(), rtol=1E-4)
+        self.assertClose(psfResult.getIxy(), psfTruth.getIxy(), rtol=1E-4)
+        self.assertFalse(result.getFlag(measBase.SdssShapeAlgorithm.PSF_SHAPE_BAD))
+
+    def testMeasureGoodPsf(self):
+        """Test that we measure shapes and record the PSF shape correctly"""
+        # Apply a varying PSF to the image to make sure we are measuring the psf at the source position
+        varyingPsf = self._computeVaryingPsf()
+        exposure, catalog = self._runMeasurementTask(psf=varyingPsf)
+        key = measBase.SdssShapeResultKey(catalog.schema["base_SdssShape"])
+        for record in catalog:
+            psfTruth = exposure.getPsf().computeShape(afwGeom.Point2D(record.getX(), record.getY()))
```

Comment:
I have absolute faith that the PSF really is varying exactly as you say it is.

But.. what if it wasn't? If we screwed up and `varyingPsf` was actually constant across the image, then this test would pass even if the original bug was still in SdssShape.

I grant it's unlikely, but could we add an extra assertion here to establish that `psfTruth` really does vary by record?